crates/polars-json/src/ndjson/deserialize.rs

```diff
@@ -14,27 +16,44 @@ use super::*;
 pub fn deserialize_iter<'a>(
     rows: impl Iterator<Item = &'a str>,
     data_type: ArrowDataType,
-    buf_size: usize,
-    count: usize,
 ) -> PolarsResult<ArrayRef> {
-    let mut buf = String::with_capacity(buf_size + count + 2);
+    let mut arr: Vec<Box<dyn Array>> = Vec::new();
+    let mut buf = String::with_capacity(std::u32::MAX as usize);
     buf.push('[');
+
+    fn _deserializer(s: &mut str, data_type: ArrowDataType) -> PolarsResult<Box<dyn Array>> {
+        // let mut buf = s.clone();
+        let slice = unsafe { s.as_bytes_mut() };
+        let out = simd_json::to_borrowed_value(slice)
+            .map_err(|e| PolarsError::ComputeError(format!("json parsing error: '{e}'").into()))?;
+        Ok(if let BorrowedValue::Array(rows) = out {
+            super::super::json::deserialize::_deserialize(&rows, data_type.clone())
+        } else {
+            unreachable!()
+        })
+    }
+
     for row in rows {
         buf.push_str(row);
-        buf.push(',')
+        buf.push(',');
+
+        if buf.len() + row.len() > (std::u32::MAX << 1) as usize {
```

Comment:
Sorry for the second review here. But I think we should use a peekable iterator.

And then 

```rust
while let Some(row) = row_iter.next() {

        let next_row_length = row_iter.peek().map(|row| row.len()).unwrap_or(0);
        if buf.len() + next_row_len > std::u32::MAX as usize {
        ..
        }

}


```

Reply:
No worries! The reviews are really helpful :D